File: Cracking the Coding Interview/8. Recursion and Dynamic Programming/8.9 N match Params.cpp
#include<iostream>
using namespace std;


void balanceParanthesis(int remaining_left,int remaining_right,char sol[],int pos){
    if(remaining_left==0){
        if(remaining_right==0){
            cout<<sol<<endl;
            return;

        }
        else{
            sol[pos]=')';
            balanceParanthesis(remaining_left,remaining_right-1,sol,pos+1);
        }
    }
    if((remaining_left<remaining_right) && (remaining_left!=0)){
        sol[pos]='(';
        balanceParanthesis(remaining_left-1,remaining_right,sol,pos+1);
        sol[pos]=')';
        balanceParanthesis(remaining_left,remaining_right-1,sol,pos+1);
    }
    if(remaining_left==remaining_right){
        sol[pos]='(';
        balanceParanthesis(remaining_left-1,remaining_right,sol,pos+1);
    }
}

int main(){
    cout<<"Enter the Value of N: ";
    int N;
    cin>>N;
    char arr[2*N];
    balanceParanthesis(N,N,arr,0);
}

File: Cracking the Coding Interview/8. Recursion and Dynamic Programming/8.9 N match Params.py
def matchParams(remain_left,remain_right,sol,pos):
    #print(remain_left,' ',remain_right)

    if(remain_left==0):
        if(remain_right==0):
            print(sol)
            return
        else:
            sol[pos]=')'
            matchParams(remain_left,remain_right-1,sol,pos+1)
    
    if(remain_left<remain_right and remain_left!=0):
        sol[pos]='('
        matchParams(remain_left-1,remain_right,sol,pos+1)
        sol[pos]=')'
        matchParams(remain_left,remain_right-1,sol,pos+1)
    if(remain_left==remain_right):
        sol[pos]='('
        matchParams(remain_left-1,remain_right,sol,pos+1)

N=3
sol = [0]*(2*N)
matchParams(N,N,sol,0)